### PR TITLE
Windows client-server fixes

### PR DESCRIPTION
**Pull request by sake92**
Only one thing in `ClientServerTests` doesn't work currently.  
Maybe someone will find a fix in the future... 😄 

I tried many things, even reimplementing the whole named pipes thing with JNA...  
But it boils down to `read/write` throwing an exception instead of returning `-1`...  
Since the read is done infinitely in a separate thread, it doesn't know when the output from server is done.  
So when a connection breaks, we should just ignore it, and not `System.exit(1);` as currently.

### CODE REVIEW

File: main/src/main/MillServerMain.scala [hunk 1]
```diff
@@ -183,33 +183,26 @@ class Server[T](lockBase: String,
...
-
-    if (Util.isWindows) {
-      // Closing Win32NamedPipeSocket can often take ~5s
-      // It seems OK to exit the client early and subsequently
-      // start up mill client again (perhaps closing the server
-      // socket helps speed up the process).
-      val t = new Thread(() => clientSocket.close())
-      t.setDaemon(true)
-      t.start()
-    } else clientSocket.close()
```
Comment (lefou) [inline main/src/main/MillServerMain.scala:195]: Where de we close the client socket now?

**Review by lefou — COMMENTED** [review]
Looks good to me. I added a question below. Does this PR really fix the server mode on Window, so that you no longer need the `--no-server` settings?

File: main/src/main/MillServerMain.scala (hunk 1 above)
Comment (sake92) [reply main/src/main/MillServerMain.scala]: Good catch, thx!  
I reverted it now. Although it doesn't do anything on windows: https://github.com/sbt/ipcsocket/blob/v1.0.1/src/main/java/org/scalasbt/ipcsocket/Win32NamedPipeSocket.java#L97

An aside: the ipcsocket library we use doesn't 100% follow the tutorial on MSDN docs.  
I tried to make an alternative implementation here https://github.com/sake92/Win32NamedPipe  
But then I realized it works even with ipcsocket as long as we ignore the exception when closing the socket in client...

**Comment by sake92** [comment]
> Looks good to me. I added a question below. Does this PR really fix the server mode on Window, so that you no longer need the `--no-server` settings?

It works 99% of the time.  
It throws this exception when you run mill on another project:
```
Exception in thread "main" java.io.IOException: Could not create lock for \\.\pipe\mill.ghrxuj6QppDw=-1_lock, error 5
        at org.scalasbt.ipcsocket.Win32NamedPipeServerSocket.<init>(Win32NamedPipeServerSocket.java:104)
```
That's because there should only be one locked pipe.  
But we give it the same pipe name, based on JDK we're running...  
Not sure how to improve on this one?

But after that it works fine. 😄 
This PR is a big improvement on status quo IMHO.

**Comment by sake92** [comment]
@lefou I think we can take inspiration from [Gradle deamon docs](https://docs.gradle.org/current/userguide/gradle_daemon.html#sec:why_is_there_more_than_one_daemon_process_on_my_machine).  
It spawns a new deamon per **JDK+(JVM properties)**.  
That seems reasonable because not all projects use same props, maybe even in tests you want different props?

Then, it doesn't seem logical to lock on `processLock/serverLock` because JDK/daemon should be global for all its clients, not per-project... (thus the exception above)

**Comment by lefou** [comment]
Thanks for this PR. I'll merge it. Nice to have mill server support on Windows.

@sake92, regarding the daemon: The mill daemon does not take multiple clients at once. Also I think we don't abstract the current working directory away in all cases, esp. when spawning sub-processes. So for having a shared daemon there is some more work to do...

**Pull request merged**